pilot/pkg/networking/core/envoyfilter/listener_patch.go

```diff
@@ -577,6 +590,62 @@ func mergeHTTPFilter(patchContext networking.EnvoyFilter_PatchContext,
 	}
 }
 
+func mergeListenerFilter(patchContext networking.EnvoyFilter_PatchContext,
+	lp *model.EnvoyFilterConfigPatchWrapper,
+	lis *listener.Listener, lisFilter *listener.ListenerFilter,
+) bool {
+	if lp.Operation != networking.EnvoyFilter_Patch_MERGE {
+		return false
+	}
+
+	if !commonConditionMatch(patchContext, lp) ||
+		!listenerMatch(lis, lp) ||
+		!listenerFilterMatch(lisFilter, lp) {
+		return false
+	}
+
+	// proto merge doesn't work well when merging two filters with ANY typed configs
+	// especially when the incoming cp.Value is a struct that could contain the json config
+	// of an ANY typed filter. So convert our filter's typed config to Struct (retaining the any
+	// typed output of json)
+	if lisFilter.GetTypedConfig() == nil {
+		// skip this op as we would possibly have to do a merge of Any with struct
+		// which doesn't seem to work well.
+		return false
+	}
+	userListenerFilter := lp.Value.(*listener.ListenerFilter)
+	var (
+		err    error
+		retVal *anypb.Any
+	)
+	if userListenerFilter.GetTypedConfig() != nil {
+		// user has any typed struct
+		// The type may not match up exactly. For example, if we use v2 internally but they use v3.
+		// Assuming they are not using deprecated/new fields, we can safely swap out the TypeUrl
+		// If we did not do this, merge.Merge below will panic (which is recovered), so even though this
+		// is not 100% reliable its better than doing nothing
+		if userListenerFilter.GetTypedConfig().TypeUrl != lisFilter.GetTypedConfig().TypeUrl {
```

Comment:
Now we use v3 every where - Do we still need this?

Reply:
it's fine to keep it as it doesn't cost too much.